Fix OperationQueue

andriydruk opened the pull request.

andriydruk committed:
Dev: fix OperationQueue

andriydruk committed:
Dev: fix wrong group enter/notify order

andriydruk committed:
Dev: replace all lock() & unlcok() with synchronized

andriydruk committed:
Dev: remove unnecessary DispatchGroups

andriydruk committed:
Dev: fix waitUntilFinished

andriydruk committed:
Dev: avoid unnecessary _runOperations in removeDependency

zayass commented on Foundation/Operation.swift:
```diff
@@ -223,25 +198,20 @@ open class BlockOperation: Operation {
-        let block = _block
-        let executionBlocks = _executionBlocks
-        lock.unlock()
+        let block = lock.synchronized { _block }
```
Unnecessary twice sync. May be omitted by storing `block` in array too

andriydruk committed:
Dev: change visibility of internal vars and funcs to fileprivate

zayass commented on Foundation/Operation.swift:
```diff
@@ -110,49 +103,56 @@ open class Operation : NSObject {
+            }
+            return true
+        }
+        if isOperationRemoved && _dependencies.count == 0 {
```
Compute this condition inside lock will be safer

zayass commented on Foundation/Operation.swift:
```diff
@@ -420,119 +348,80 @@ open class OperationQueue: NSObject {
-        if let op = _dequeueOperation() {
-            if !op.isCancelled {
-                op._waitUntilReady()
+    private var _operationCount = 0
```
Internal naming is confusing `operationCount` means "all operations" but `_operationCount` means "running operations"

andriydruk committed:
Dev: add asserts for addDependency

andriydruk committed:
Dev: add support for custom isReady and macOS 10.6 cancel behaviour

> In macOS 10.6 and later, if you cancel an operation while it is waiting on the completion of one or more dependent operations, those dependencies are thereafter ignored and the value of this property is updated to reflect that it is now ready to run. This behavior gives an operation queue the chance to flush cancelled operations out of its queue more quickly.

andriydruk committed:
Cleanup

zayass commented on Foundation/Operation.swift:
```diff
@@ -347,6 +344,8 @@ open class OperationQueue: NSObject {
     }
     fileprivate let queueGroup = DispatchGroup()
     fileprivate var _operations = _OperationList()
+    fileprivate var _runningOperationCount = 0
```
why not private?
